File: website_backup/src/lib/noaaClient.ts
import type { LiveObservation, NoaaObservationResponse } from "../types/station";

const NOAA_API_BASE_URL = "https://api.weather.gov";

export class NoaaClientError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NoaaClientError";
  }
}

function normalizeNumber(value: unknown): number | null {
  if (typeof value !== "number") {
    return null;
  }

  if (!Number.isFinite(value)) {
    return null;
  }

  return value;
}

export async function fetchLatestObservation(
  stationId: string
): Promise<LiveObservation> {
  const normalizedStationId = stationId.trim().toUpperCase();

  if (!normalizedStationId) {
    throw new NoaaClientError("Station ID is required.");
  }

  const url = `${NOAA_API_BASE_URL}/stations/${normalizedStationId}/observations/latest`;

  const response = await fetch(url, {
    headers: {
      Accept: "application/geo+json",
      "User-Agent":
        "renewable-energy-forecasting-pipeline-website (portfolio project)",
    },
    cache: "no-store",
  });

  if (!response.ok) {
    throw new NoaaClientError(
      `NOAA request failed for ${normalizedStationId}: ${response.status} ${response.statusText}`
    );
  }

  const payload = (await response.json()) as NoaaObservationResponse;

  const properties = payload.properties;

  if (!properties) {
    throw new NoaaClientError(
      `NOAA response for ${normalizedStationId} did not include observation properties.`
    );
  }

  return {
    stationId: normalizedStationId,
    timestamp: properties.timestamp ?? null,
    windSpeedMs: normalizeNumber(properties.windSpeed?.value),
    windDirectionDeg: normalizeNumber(properties.windDirection?.value),
    temperatureC: normalizeNumber(properties.temperature?.value),
    rawSource: "NOAA_NWS",
  };
}

File: model_service/app/noaa_client.py
from __future__ import annotations

from datetime import datetime, timezone
import httpx


class NOAAClient:
    def __init__(self, user_agent: str = "wind-energy-forecasting-platform/1.0"):
        self.base_url = "https://api.weather.gov"
        self.headers = {
            "User-Agent": user_agent,
            "Accept": "application/geo+json",
        }

    async def get_latest_observation(self, station_id: str) -> dict:
        station_id = station_id.upper().strip()
        url = f"{self.base_url}/stations/{station_id}/observations/latest"

        async with httpx.AsyncClient(timeout=15) as client:
            response = await client.get(url, headers=self.headers)
            response.raise_for_status()
            payload = response.json()

        props = payload.get("properties", {})

        wind_speed_ms = self._extract_value(props.get("windSpeed"))
        wind_direction_deg = self._extract_value(props.get("windDirection"))
        temperature_c = self._extract_value(props.get("temperature"))
        timestamp = props.get("timestamp")

        return {
            "station_id": station_id,
            "timestamp": timestamp,
            "wind_speed_ms": wind_speed_ms,
            "wind_direction_deg": wind_direction_deg,
            "temperature_c": temperature_c,
            "observation_age_minutes": self._age_minutes(timestamp),
            "raw_source": "NOAA/NWS latest observation",
        }

    @staticmethod
    def _extract_value(field: dict | None) -> float | None:
        if not isinstance(field, dict):
            return None
        value = field.get("value")
        return None if value is None else float(value)

    @staticmethod
    def _age_minutes(timestamp: str | None) -> float | None:
        if not timestamp:
            return None
        try:
            observed = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
            now = datetime.now(timezone.utc)
            return round((now - observed).total_seconds() / 60.0, 1)
        except Exception:
            return None
